ot/bregman.py

```diff
@@ -1787,3 +1789,287 @@ def empirical_sinkhorn_divergence(X_s, X_t, reg, a=None, b=None, metric='sqeucli
+    ns, nt = M.shape
+
+    # calculate the Gibbs kernel
+    K = np.empty_like(M)
+    np.divide(M, - reg, out=K)
+    np.exp(K, out=K)
+
+    def projection(u, epsilon):
+        u[np.where(u <= epsilon)] = epsilon
+        return u
+
+    # ----------------------------------------------------------------------------------------------------------------#
+    #                                          Step 1: Screening Pre-processing                                       #
+    # ----------------------------------------------------------------------------------------------------------------#
+
+    if ns_budget == ns and nt_budget == nt:
+        # full number of budget points (ns, nt) = (ns_budget, nt_budget)
+        I = list(range(ns))
+        J = list(range(nt))
+        epsilon = 0.0  
+        kappa = 1.0  
+
+        cst_u = 0.
+        cst_v = 0.
+
+        bounds_u = [(0.0, np.inf)] * ns
+        bounds_v = [(0.0, np.inf)] * nt
+        
+        a_I = a
+        b_J = b
+        K_min = K.min()
+        K_IJ = K
+        K_IJc = []
+        K_IcJ = []
+
+        vec_eps_IJc = np.zeros(nt)
+        vec_eps_IcJ = np.zeros(ns)
+
+    else:
+        # sum of rows and columns of K
+        K_sum_cols = K.sum(axis=1)
+        K_sum_rows = K.sum(axis=0)
+
+        if uniform:
+            if ns / ns_budget < 4:
+                aK_sort = np.sort(K_sum_cols)
+                epsilon_u_square = a[0] / aK_sort[ns_budget - 1]
+            else:
+                aK_sort = bottleneck.partition(K_sum_cols, ns_budget - 1)[ns_budget - 1]
+                epsilon_u_square = a[0] / aK_sort
+
+            if nt / ns_budget < 4:
+                bK_sort = np.sort(K_sum_rows)
+                epsilon_v_square = b[0] / bK_sort[ns_budget - 1]
+            else:
+                bK_sort = bottleneck.partition(K_sum_rows, nt_budget - 1)[nt_budget - 1]
+                epsilon_v_square = b[0] / bK_sort
+        else:
+            aK = a / K_sum_cols
+            bK = b / K_sum_rows
+
+            aK_sort = np.sort(aK)[::-1]
+            epsilon_u_square = aK_sort[ns_budget - 1]
+
+            bK_sort = np.sort(bK)[::-1]
+            epsilon_v_square = bK_sort[ns_budget - 1]
+
+        # active sets I and J (see Proposition .. in [1])
+        I = np.where(a >= epsilon_u_square * K_sum_cols)[0].tolist()
+        J = np.where(b >= epsilon_v_square * K_sum_rows)[0].tolist()
+
+        if len(I) != ns_budget:
+            if uniform:
+                aK = a / K_sum_cols
+                aK_sort = np.sort(aK)[::-1]
+            epsilon_u_square = aK_sort[ns_budget - 1:ns_budget + 1].mean()
+            I = np.where(a >= epsilon_u_square * K_sum_cols)[0].tolist()
+
+        if len(J) != nt_budget:
+            if uniform:
+                bK = b / K_sum_rows
+                bK_sort = np.sort(bK)[::-1]
+            epsilon_v_square = bK_sort[ns_budget - 1:ns_budget + 1].mean()
+            J = np.where(b >= epsilon_v_square * K_sum_rows)[0].tolist()
```

Comment:
could you stay with array of integer?

i would be suprised if np.ix_ couldn't handle those.